### ISSUE TITLE
OOM within docker

### ISSUE THREAD

**Issue by keithchew** [ledgerwatch/erigon#6646]
#### System information

Erigon version: `./erigon --version`
v2.35.2

OS & Version: Windows/Linux/OSX
Ubuntu 20.04 (docker-ce 20.10.22)
OS: 64GB RAM
docker container max memory: 32GB

#### Expected behaviour
Not to observe unbounded memory increase.

#### Actual behaviour
Memory keeps increasing until OOM.

#### Steps to reproduce the behaviour
run erigon with:
--pprof --chain=mainnet --batchSize=8mb --torrent.conns.perfile=4 --torrent.download.rate=100mb
GOGC=50
GOMEMLIMIT=6000MiB
GODEBUG=gctrace=1

After catching up with latest block, memory is stable at 12GB. gc trace confirms erigon itself only uses 2.9GB.

gc 321 @7184.801s 0%: 0.53+151+0.020 ms clock, 12+26/908/1809+0.48 ms cpu, 4029->4110->2946 MB, 4472 MB goal, 43 MB stacks, 2 MB globals, 24 P

Rest of RAM is assumed to be used by OS page cache.

While erigon is running, memory consumption grows steadily, until OOM kicks in at 32GB and kills the erigon process. 

Taking a wild guess, could this be caused by libmdbx thinking there is 64GB memory (OS RAM), but in actual fact the docker container is limited at 32GB? Since we can limit GO with the GOMEMLIMIT flag, is there something we can pass to libmdbx to specify the max memory to use?

**Comment by AskAlexSharov** [comment]
to understand where Erigon wasted RAM can do:
add flag `--pprof`
`go tool pprof -inuse_space -png http://127.0.0.1:6060/debug/pprof/heap > mem.png`
show this file

**Comment by keithchew** [comment]
Hi Alex

Go prof diagram shows Erigon is not wasting RAM, as noted in the GC log above, ie only 2.9GB used. It is the RAM outside of GO that is slowly growing.

**Comment by AskAlexSharov** [comment]
OOM can’t kill Erigon for external memory, because that memory owned by OS (even can survive Erigon’s reboot). See https://github.com/ledgerwatch/erigon#htop-shows-incorrect-memory-usage

**Comment by AskAlexSharov** [comment]
@keithchew do you have much total RAM on machine? I guess Erigon does some estimations on TotalRam but not on GOMEMLIMIT

**Comment by AskAlexSharov** [comment]
if yes, try slightly reduce GOMAXPROC

**Comment by keithchew** [comment]
Hi Alex

Total RAM on machine is 64GB. 

OOM is killing erigon because the memory reported by docker stats (and confirmed using cgroups) reached 32GB, even though GO is reporting only 2.9GB is used.  So, my assumption was there is another 29.1GB used by the container but reported in GO's pprof/gc trace, is this even possible?

OK, will read up on GOMAXPROC and keep us posted.

**Comment by zqt996** [comment]
Use docker to limit memory limits

**Comment by keithchew** [comment]
As per the description above, memory is already limited in the docker container to 32GB, and it is not a good idea to disable OOM, otherwise it can trigger bigger issues when memory is exhausted.

**Comment by emgeee** [comment]
just to chime in here -- I'm seeing something similar with Kubernetes on both versions 2.35 and 2.36

**Comment by keithchew** [comment]
Hi Alex

I have limited docker container to 20GB instead of 32GB (to see if it will trigger an OOM faster) and have set GOMAXPROC=4 instead of default 24.

I can observe a slowdown, since erigon only has 4 CPUs for parallelism. But RAM is stable at 16GB! Thank you very much for the recommendation, will be able to tweak GOMAXPROC (and other params) to match load and available RAM.

Will keep monitoring and post here if there are further issues.

**Comment by AskAlexSharov** [comment]
Estimation logic is here: https://github.com/ledgerwatch/erigon/blob/0eebd61ab8cee7eb7590076975700cdc54b68c43/eth/ethconfig/estimate/esitmated_ram.go#L14

maybe you can advise us something

**Comment by keithchew** [comment]
Hi Alex

Thank you very much for the reference to that code. 

Ah ha, memory.TotalMemory() will return the physical RAM available, without any visibility to cgroups memory limit within docker.

The code in estimated_ram might already be correct, and maybe we could have an optional runtime param (say MAX_MEM) which allows us to pass in the value that replaces memory.TotalMemory(). This should work well for users running erigon in docker containers. The current workaround is to limit it via GOMAXPROC.

**Comment by AskAlexSharov** [comment]
No, it’s out of Erigon project scope. Just some lib must check cgroups limits under the hood: https://stackoverflow.com/a/42188649

**Comment by keithchew** [comment]
Yes it would be better if done under the hood. At least we have a workaround in the meantime.

**Comment by AskAlexSharov** [comment]
This one does: https://github.com/shirou/gopsutil/blob/301930f9bdd0d576ebb7940a13c6b9242df4e370/docker/docker_linux.go#L259

**Comment by fafrd** [comment]
@AskAlexSharov The library shirou/gopsutil doesn't seem like the right choice. On my test system (debian 11, docker, ubuntu 20.04 guest) the cgroup info does not show up in the expected directory:
code:
```go
    cgmem, err := docker.CgroupMemDocker(hostname)
    if err != nil {
        fmt.Println(err)
        os.Exit(1)
    }
    fmt.Printf("CgroupMemStat: %+v\n", cgmem)
```
output:
```bash
(ins) ~/erigon-testing/erigon $ docker run --memory 256m -it image /bin/bash
root@71173d45655b:/# /usr/local/bin/erigon
open /sys/fs/cgroup/memory/system.slice/docker-71173d45655b.scope/memory.stat: no such file or directory
```
The directory `/sys/fs/cgroup/memory` doesn't even exist, on the host or the guest... 

However, I do see the correct memory limit in this file:
```bash
root@71173d45655b:/# cat /sys/fs/cgroup/memory.max
268435456
```

Turns out this is the difference between cgroup and cgroup2. cgroup2 was introduced in linux 4.5 (in 2016). We should presumably only be targeting cgroup2 support at this point. SO, i'm going to try to resolve this by reading the file `/sys/fs/cgroup/memory.max`, using that if it exists, else falling back to `memory.TotalMemory()`.

(I opened up an issue with gopsutil: https://github.com/shirou/gopsutil/issues/1416)

**Comment by keithchew** [comment]
on my machine, the mem info (from guest) is in:
/sys/fs/cgroup/memory/memory.max_usage_in_bytes

and from host:
/sys/fs/cgroup/memory/docker/<container id>/memory.max_usage_in_bytes

So, the gopsutil library will work. This machine is in Ubuntu 20.04, and although using 5.15 kernel, the default installation is configured to use cgroup1.

I would recommend adding cgroup2 support into the gopsutil library instead of rolling out your own in Erigon. This will enable the library to return the correct mem usage for both cgroup and cgroup2.

**Comment by tdiesler** [comment]
Is it really necessary for erigon to consume > 32G on initial sync? On a machine with 32G total Ram it crashed on me twice. I assumed when setting a docker `--memory` limit (e.g. 28G) it would respect that and adjust its memory consumption accordingly.

Anyway, a hard oom crash seem wrong when running in docker

PS: v2.40.1

**Comment by AskAlexSharov** [comment]
@tdiesler workaround it by reduce GOMAXPROCS

**Comment by AskAlexSharov** [comment]
let's keep this issue alive

**Comment by elee1766** [comment]
> @AskAlexSharov The library shirou/gopsutil doesn't seem like the right choice. On my test system (debian 11, docker, ubuntu 20.04 guest) the cgroup info does not show up in the expected directory:
> code:
> ```go
>     cgmem, err := docker.CgroupMemDocker(hostname)
>     if err != nil {
>         fmt.Println(err)
>         os.Exit(1)
>     }
>     fmt.Printf("CgroupMemStat: %+v\n", cgmem)
> ```
> output:
> ```bash
> (ins) ~/erigon-testing/erigon $ docker run --memory 256m -it image /bin/bash
> root@71173d45655b:/# /usr/local/bin/erigon
> open /sys/fs/cgroup/memory/system.slice/docker-71173d45655b.scope/memory.stat: no such file or directory
> ```
> The directory `/sys/fs/cgroup/memory` doesn't even exist, on the host or the guest... 
> 
> However, I do see the correct memory limit in this file:
> ```bash
> root@71173d45655b:/# cat /sys/fs/cgroup/memory.max
> 268435456
> ```
> 
> Turns out this is the difference between cgroup and cgroup2. ~cgroup2 was introduced in linux 4.5 (in 2016). So, i'm going to try to resolve this by reading the file `/sys/fs/cgroup/memory.max`, using that if it exists, else falling back to `memory.TotalMemory()`.~
> 
> EDIT- actually, it started getting adopted by distros around 2021, so relatively recently: https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md
> 
> (I opened up an issue with gopsutil: https://github.com/shirou/gopsutil/issues/1416) 
> 

hi, try this package for interacting with cgroups, I have had success with it (it's also what k8s uses I believe)

https://github.com/containerd/cgroups

**Comment by elee1766** [comment]
hi, chiming in


this is not a bug in erigon, it is a "feature" of cgroups. I say feature because it's really a bug.


cgroups will count page cache memory usage. the first cgroup to touch a page in page cache, that page will count for the memory usage of the cgroup. however cgroup v1 has no good tools for automatic reclaim of such


from kernel docs for cgroup v1
```
2.3 Shared Page Accounting

Shared pages are accounted on the basis of the first touch approach. The
cgroup that first touches a page is accounted for the page. The principle
behind this approach is that a cgroup that aggressively uses a shared
page will eventually get charged for it (once it is uncharged from
the cgroup that brought it in -- this will happen on memory pressure).

But see section 8.2: when moving a task to another cgroup, its pages may
be recharged to the new cgroup, if move_charge_at_immigrate has been chosen.

Exception: If CONFIG_MEMCG_SWAP is not used.
When you do swapoff and make swapped-out pages of shmem(tmpfs) to
be backed into memory in force, charges for pages are accounted against the
caller of swapoff rather than the users of shmem
```

this causes applications with high page cache usage to constantly get oom killed when running in such environment 

a related issue in kubernetes repo

https://github.com/kubernetes/kubernetes/issues/43916


supposedly cgroup V2 has page reclaim ? but I don't know if it works. many people r still running v1 even if they have support unified in kernel


this is possible fixable in erigon - need to check cgroup usage periodically and tell kernel from within erigon to dispose it's page cache periodically, but I have not gone too deep into it. my solution for now - have more ram when running on cgroup based environment

**Comment by AskAlexSharov** [comment]
@keithchew see https://github.com/ledgerwatch/erigon#htop-shows-incorrect-memory-usage

**Comment by hitchhooker** [comment]
its not only docker but also lx-containers. some flag for manually setting ram amounts would be awesome workaround solution. 
been running erigon in containers now for half a year without able to keep in sync after initial sync and it seems this is the most likely reason for it.

**Comment by alecalve** [comment]
I am running in that issue trying to sync Polygon on a beefy machine where Erigon is co-located with other nodes and only allocated a subset of the available memory (using Docker).

Could a possible simple fix be to allow overriding via the config the maximum number of workers for specific stages?

**Comment by AskAlexSharov** [comment]
@alecalve workaround is reduce a bit GOMAXPROCS during initial sync. also can add GOGC=50

**Comment by yperbasis** [comment]
Fixed by PR #8632
